From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Tue, 15 Apr 2025 08:13:18 +0000
Subject: [PATCH 096/243] NVIDIA: SAUCE: virt: add sanity checks for ivc info's
 metadata

Jira ESLC-8494
Bug 5226392

Change-Id: I9fdad5f5b65acb637ea889becc2c637f1747afbd
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3427616
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit aa245d4b56f521afa32f83799df15e80e8d79a90)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index 5261bde1be76c..38863277b0c6d 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -517,6 +517,19 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 		return -ENOMEM;
 	}
 
+	if ((hvd->info->nr_queues == 0U) || (hvd->info->nr_queues > PCT_MAX_NUM_IVC_QUEUES)) {
+		ERR("Invalid nr_queues: %u\n", hvd->info->nr_queues);
+		return -EINVAL;
+	}
+	if (hvd->info->nr_areas > MAX_NUM_GUESTS) {
+		ERR("Invalid nr_areas: %u\n", hvd->info->nr_areas);
+		return -EINVAL;
+	}
+	if (hvd->info->nr_mempools > PCT_MAX_NUM_MEMPOOLS) {
+		ERR("Invalid nr_mempools: %u\n", hvd->info->nr_mempools);
+		return -EINVAL;
+	}
+
 	/*
 	 *  Map IVC Trap MMIO Notification region
 	 */
diff --git a/include/soc/tegra/virt/tegra_hv.h b/include/soc/tegra/virt/tegra_hv.h
index c64db45e90efd..00c4fcb0e5db0 100644
--- a/include/soc/tegra/virt/tegra_hv.h
+++ b/include/soc/tegra/virt/tegra_hv.h
@@ -12,6 +12,13 @@
 
 #define IVC_INFO_PAGE_SIZE 65536
 
+/** @brief Maximum Guest VM count */
+#define MAX_NUM_GUESTS		16U
+/** @brief The maximum number of IVC queues supported by the PCT. */
+#define PCT_MAX_NUM_IVC_QUEUES	512U
+/** @brief The maximum number of mempools supported by the PCT. */
+#define PCT_MAX_NUM_MEMPOOLS	120U
+
 const struct ivc_info_page *tegra_hv_get_ivc_info(void);
 int tegra_hv_get_vmid(void);
 
-- 
2.43.0

